### PR TITLE
Split train- and val progress into separate bars

### LINKED ISSUE #12623: Separate training/validation progress bars
## 🚀 Feature

Currently, there is a global progress bar for training + validation, and a validation progress bar during validation. It would be very helpful to have a separate training-only progress bar.

### Motivation

1. A global progress bar does not accurately estimate the time remaining because training and validation have different speed. A training-only progress bar at least gives us an unbiased estimate of the time remaining for the training stage.
2. The n_total in the global training bar considers both training and validation batches. Having a training-only progress bar that displays the number of training batches is helpful for calculating gradient accumulation, distributed training nodes, etc.

### Pitch

An option to either replace the global progress bar with a training-only progress bar, or to add a training-only progress bar.

cc @borda @kaushikb11 @rohitgr7 @awaelchli @carmocca

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #12623

A long pending request from the community is an improved progress bar. So far, the training and validation counts were always added together in one progress bar. This has the drawback that more often than not, the time estimate for completion is wrong, because the validation loop typically runs much faster than the regular training loop. Furthermore, users were often confused about the total iteration count, because they didn't realize that the total number was train + val batches. See #6735, #1121, #6406

This PR splits the bars into two separate ones. The validation bar appears and disappears as needed as before. The PR introduces the change for both TQDM and Rich.

Video recording with 10 train batches 5 val batches, validation every 2 train batches:

TQDM:

https://user-images.githubusercontent.com/5495193/217672742-0d188992-6a72-45c1-b05f-0c3bf32924d6.mov

Rich:

https://user-images.githubusercontent.com/5495193/217672887-7780de99-83c4-4e21-8bc1-804e35998a18.mov

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @justusschock @awaelchli

### CODE REVIEW

File: src/lightning/pytorch/callbacks/progress/rich_progress.py [hunk 1]
```diff
@@ -362,7 +362,7 @@ def on_sanity_check_end(self, trainer: "pl.Trainer", pl_module: "pl.LightningMod
     def on_train_epoch_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
         if self.is_disabled:
             return
-        total_batches = self.total_batches_current_epoch
+        total_batches = self.total_train_batches
         train_description = self._get_train_description(trainer.current_epoch)
 
         if self.main_progress_bar_id is not None and self._leave:
```
Comment (carmocca) [inline src/lightning/pytorch/callbacks/progress/rich_progress.py:368]: Maybe we should rename it to `train_progress_bar` now for both files. Maybe in a follow-up

File: src/lightning/pytorch/callbacks/progress/rich_progress.py [hunk 2]
```diff
@@ -491,9 +491,6 @@ def on_validation_batch_end(
         if trainer.sanity_checking:
             self._update(self.val_sanity_progress_bar_id, self.val_batch_idx)
         elif self.val_progress_bar_id is not None:
-            # check to see if we should update the main training progress bar
-            if self.main_progress_bar_id is not None:
-                self._update(self.main_progress_bar_id, self.train_batch_idx + self._val_processed)
             self._update(self.val_progress_bar_id, self.val_batch_idx)
         self.refresh()
```
Comment (carmocca) [inline src/lightning/pytorch/callbacks/progress/rich_progress.py:496]: Do we need the code in L531 (`_get_train_description`)?

File: src/lightning/pytorch/callbacks/progress/tqdm_progress.py [hunk 3]
```diff
@@ -252,13 +252,12 @@ def on_train_start(self, *_: Any) -> None:
         self.main_progress_bar = self.init_train_tqdm()
```
Comment (carmocca) [inline src/lightning/pytorch/callbacks/progress/tqdm_progress.py:252]: L66 (docstring) needs an update

File: src/lightning/pytorch/callbacks/progress/rich_progress.py (hunk 2 above)
Comment (awaelchli) [reply src/lightning/pytorch/callbacks/progress/rich_progress.py:496]: Yes I think so. It is about the validation bar below the train bar. I don't think it is related to this PR. I'm not sure though.

File: src/lightning/pytorch/callbacks/progress/tqdm_progress.py (hunk 3 above)
Comment (awaelchli) [reply src/lightning/pytorch/callbacks/progress/tqdm_progress.py:252]: Yes, updated docs where I found references

File: src/lightning/pytorch/callbacks/progress/rich_progress.py (hunk 1 above)
Comment (awaelchli) [reply src/lightning/pytorch/callbacks/progress/rich_progress.py]: Yes. However, it could also be that there are few users who made a custom bar relying on this public attribute.

File: docs/source-pytorch/common/progress_bar.rst [hunk 4]
```diff
@@ -22,7 +22,7 @@ The :class:`~pytorch_lightning.callbacks.TQDMProgressBar` uses the `tqdm <https:
 It prints to ``stdout`` and shows up to four different bars:
 
 - **sanity check progress:** the progress during the sanity check run
-- **main progress:** shows training + validation progress combined. It also accounts for multiple validation runs during training when :paramref:`~pytorch_lightning.trainer.trainer.Trainer.val_check_interval` is used.
+- **main progress:** shows the training progress. It will pause if validation starts and will resume when it ends, and also accounts for multiple validation runs during training when :paramref:`~pytorch_lightning.trainer.trainer.Trainer.val_check_interval` is used.
```
Comment (carmocca) [inline docs/source-pytorch/common/progress_bar.rst:25]: Did you keep "main" here on purpose?

File: src/lightning/pytorch/CHANGELOG.md [hunk 5]
```diff
@@ -58,6 +58,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - The strategy selected by `accelerator="hpu"` now defaults to `find_unused_parameters=False` ([#16611](https://github.com/Lightning-AI/lightning/pull/16611))
 
+- The main progress bar displayed during training no longer includes the combined progress for validation ([#16695](https://github.com/Lightning-AI/lightning/pull/16695))
```
Comment (carmocca) [inline src/lightning/pytorch/CHANGELOG.md:61]: Can you mention that main was renamed?

File: src/lightning/pytorch/callbacks/progress/tqdm_progress.py [hunk 6]
```diff
@@ -63,8 +63,8 @@ class TQDMProgressBar(ProgressBarBase):
     :mod:`tqdm` package and shows up to four different bars:
 
         - **sanity check progress:** the progress during the sanity check run
-        - **main progress:** shows training + validation progress combined. It also accounts for
-          multiple validation runs during training when
+        - **main progress:** shows the training progress. It will pause if validation starts and will resume
```
Comment (carmocca) [inline src/lightning/pytorch/callbacks/progress/tqdm_progress.py:66]: Same question

**Review by carmocca — APPROVED** [review]

File: docs/source-pytorch/common/progress_bar.rst (hunk 4 above)
Comment (awaelchli) [reply docs/source-pytorch/common/progress_bar.rst]: was missed because of missing underscore

File: src/lightning/pytorch/CHANGELOG.md (hunk 5 above)
Comment (awaelchli) [reply src/lightning/pytorch/CHANGELOG.md:61]: yes

**Review by justusschock — APPROVED** [review]

**Review by Borda — APPROVED** [review]
Just curious do we have also a bar for testing and predictions?

**Comment by awaelchli** [comment]
@Borda These are already there and remain unchanged. They were never coupled to the "main" bar, so no changes there.
